### PR TITLE
VMware: Use existing disks in vmware_guest_disk module

### PR DESCRIPTION
**Pull request by jeking3**
##### SUMMARY

Extended the changes from #45953 (existing disks in vmware_guest) to vmware_guest_disk.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

vmware_guest_disk

### CODE REVIEW

**Comment by jeking3** [comment]
Once #45953 is merged, I will refresh this.

**Comment by Akasurde** [comment]
@jeking3 Should we merge this first as it will consolidate the API in `module_utils/vmware.py` ? Let me know.

File: lib/ansible/module_utils/vmware.py [hunk 1]
```diff
@@ -1173,3 +1174,64 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
...
+        3. vmdk_filename: The "vm_name.vmdk" portion of the string (os.path.basename equivalent)
+        4. vmdk_folder: The "/path/to/" portion of the string (os.path.dirname equivalent)
+        """
+        try:
+            datastore_name = re.match(r'^\[(.*?)\]', vmdk_path, re.DOTALL).groups()[0]
+            vmdk_fullpath = re.match(r'\[.*?\] (.*)$', vmdk_path).groups()[0]
+            vmdk_filename = os.path.basename(vmdk_fullpath)
+            vmdk_folder = os.path.dirname(vmdk_fullpath)
+        except (IndexError, AttributeError) as e:
+            self.module.fail_json("vmdk full path '{0}' is not in the canonical form '[datastore_name] /path/to/vm_name.vmdk'".format(vmdk_path))
```
Comment (Akasurde) [inline lib/ansible/module_utils/vmware.py]:
```suggestion
            self.module.fail_json(msg="vmdk full path '{0}' is not in the canonical form '[datastore_name] /path/to/vm_name.vmdk'".format(vmdk_path))
```

File: lib/ansible/module_utils/vmware.py [hunk 2]
```diff
@@ -1173,3 +1174,64 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
...
+        while search_res.info.state not in [vim.TaskInfo.State.success,
+                                            vim.TaskInfo.State.error]:
+            time.sleep(1)
+        if search_res.info.result is None:
+            return None
+        target_folder_path = "[" + datastore_name + "]" + " " + vmdk_folder + '/'
+        for result in search_res.info.result:
+            for f in getattr(result, 'file'):
+                if f.path == vmdk_filename and result.folderPath == target_folder_path:
+                   return f
```
Comment (Akasurde) [inline lib/ansible/module_utils/vmware.py]: need spaces with multiple of 4

File: lib/ansible/modules/cloud/vmware/vmware_guest_disk.py [hunk 3]
```diff
@@ -75,6 +75,9 @@
      - '     - C(thick) thick disk'
      - '     Default: C(thick) thick disk, no eagerzero.'
      - ' - C(datastore) (string): Name of datastore or datastore cluster to be used for the disk.'
+     - ' - C(filename) (string): Existing disk image to use.  Must already exist on the datastore.'
+     - '   Value is of form C(path/to/file.vmdk).'
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_guest_disk.py:79]: Add a note that - `path should not start with '/'`

File: lib/ansible/module_utils/vmware.py [hunk 4]
```diff
@@ -1173,3 +1174,64 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
...
+            vmdk_filename = os.path.basename(vmdk_fullpath)
+            vmdk_folder = os.path.dirname(vmdk_fullpath)
+        except (IndexError, AttributeError) as e:
+            self.module.fail_json("vmdk full path '{0}' is not in the canonical form '[datastore_name] /path/to/vm_name.vmdk'".format(vmdk_path))
+        return (datastore_name, vmdk_fullpath, vmdk_filename, vmdk_folder)
+
+    def find_vmdk(self, vmdk_path):
+        """
+        Takes a vsphere datastore path in the format
+             [datastore_name] /path/to/file.vmdk
```
Comment (Akasurde) [inline lib/ansible/module_utils/vmware.py]:
```suggestion
             [datastore_name] path/to/file.vmdk
```

File: lib/ansible/module_utils/vmware.py [hunk 5]
```diff
@@ -1173,3 +1174,64 @@ def find_datastore_cluster_by_name(self, datastore_cluster_name):
             if dsc.name == datastore_cluster_name:
                 return dsc
         return None
+
+    def vmdk_disk_path_split(self, vmdk_path):
+        """
+        Takes a string in the format
+             [datastore_name] /path/to/vm_name.vmdk
```
Comment (Akasurde) [inline lib/ansible/module_utils/vmware.py]:
```suggestion
             [datastore_name] path/to/vm_name.vmdk
```

**Comment by Akasurde** [comment]
@jeking3 Any news ? I am using the `find_vmdk` API in https://github.com/ansible/ansible/pull/50923

**Comment by jeking3** [comment]
Last time I looked at this PR I was waiting for an upstream PR to be merged, noted in the description.  That hasn't been merged, so this is on hold until that happens.

**Comment by Akasurde** [comment]
@jeking3 Since your PR includes APIs in `module_utils/vmware.py` it would be good to merge this first. Then #50923 and #45953 can be merged. What do you think ? 

@mrmagooey Are you OK merging this PR before #45953 ?

**Comment by jeking3** [comment]
I thought some work would need to be done after #45953 was committed to refactor but it's up to you if you want to change the commit order and resolution.

**Comment by jeking3** [comment]
Let me rebase this and see what's left...

**Comment by jeking3** [comment]
This could use an integration test; I didn't know how to make them for Ansible at the time I submitted it.

**Comment by Akasurde** [comment]
@jeking3 Could you please review this ? Thanks.

**Comment by jeking3** [comment]
Biggest issue is that it's missing an integration test, otherwise ok.

**Comment by jeking3** [comment]
@Akasurde this looks okay but I don't have an environment I can test it in at the moment.

**Pull request closed without merging**